Tests/PackageGraphTests/PackageGraphTests.swift

```diff
@@ -1255,4 +1263,551 @@ class PackageGraphTests: XCTestCase {
+            "/Bar/Sources/Bar/bar.swift"
+        )
+
+        let diagnostics = DiagnosticsEngine()
+        _ = try loadPackageGraph(fs: fs, diagnostics: diagnostics,
+            manifests: [
+                Manifest.createManifest(
+                    name: "Foo",
+                    path: "/Foo",
+                    packageKind: .root,
+                    packageLocation: "/Foo",
+                    v: .v5_2,
+                    dependencies: [
+                        .scm(location: "/Bar", requirement: .upToNextMajor(from: "1.0.0")),
+                    ],
+                    targets: [
+                        TargetDescription(name: "Foo", dependencies: ["Bar"]),
+                    ]),
+                Manifest.createManifest(
+                    name: "Bar",
+                    path: "/Bar",
+                    packageKind: .local,
+                    packageLocation: "/Bar",
+                    v: .v5_2,
+                    products: [
+                        ProductDescription(name: "Bar", type: .library(.automatic), targets: ["Bar"])
+                    ],
+                    targets: [
+                        TargetDescription(name: "Bar"),
+                    ]),
+            ]
+        )
+
+        XCTAssert(diagnostics.diagnostics.isEmpty, "\(diagnostics.diagnostics)")
+    }
+
+    func testTargetDependencies_Post52_UnknownProduct() throws {
+        let fs = InMemoryFileSystem(emptyFiles:
+            "/Foo/Sources/Foo/foo.swift",
+            "/Bar/Sources/Bar/bar.swift"
+        )
+
+        let diagnostics = DiagnosticsEngine()
+        _ = try loadPackageGraph(fs: fs, diagnostics: diagnostics,
+            manifests: [
+                Manifest.createManifest(
+                    name: "Foo",
+                    path: "/Foo",
+                    packageKind: .root,
+                    packageLocation: "/Foo",
+                    v: .v5_2,
+                    dependencies: [
+                        .scm(location: "/Bar", requirement: .upToNextMajor(from: "1.0.0")),
+                    ],
+                    targets: [
+                        TargetDescription(name: "Foo", dependencies: ["Unknown"]),
+                    ]),
+                Manifest.createManifest(
+                    name: "Bar",
+                    path: "/Bar",
+                    packageKind: .local,
+                    packageLocation: "/Bar",
+                    v: .v5_2,
+                    products: [
+                        ProductDescription(name: "Bar", type: .library(.automatic), targets: ["Bar"])
+                    ],
+                    targets: [
+                        TargetDescription(name: "Bar"),
+                    ]),
+            ]
+        )
+
+        DiagnosticsEngineTester(diagnostics, ignoreNotes: true) { result in
+            result.check(
+                diagnostic: """
+                    product 'Unknown' not found. it is required by package 'Foo' target 'Foo'.
```

Comment:
Not specific to this test, but does this kind of message now become ambiguous? It feels like conceptually we could now be referring to `Package.name` or to the last path component of the URL since we are treating them equally, even when they differ. I wonder if for remote packages, we should instead emit the URL here.